File: 02/java/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import static java.util.stream.Collectors.toList;

class Main {
    enum Direction {
        Vertical, Horizontal;
    }

    record Command(Direction direction, int val) {
        static Command parse(String s) {
            var split = s.split(" ");
            var direction = split[0];
            int val = Integer.parseInt(split[1]);
            return switch (direction) {
                case "up" -> new Command(Direction.Vertical, -val);
                case "down" -> new Command(Direction.Vertical, val);
                case "forward" -> new Command(Direction.Horizontal, val);
                default -> throw new IllegalArgumentException("bad command string: " + s);
            };
        }
    };

    record Position(int x, int depth) {
        static final Position ZERO = new Position(0, 0);

        Position update(Command cmd) {
            return switch (cmd.direction) {
                case Vertical -> new Position(x, depth + cmd.val);
                case Horizontal -> new Position(x + cmd.val, depth);
            };
        }

        int product() {
            return x * depth;
        }
    }

    record PositionWithAim(Position pos, int aim) {
        PositionWithAim update(Command c) {
            return switch (c.direction) {
                case Vertical -> new PositionWithAim(pos, aim + c.val);
                case Horizontal -> new PositionWithAim(new Position(pos.x + c.val, pos.depth + aim * c.val), aim);
            };
        }
    }

    static int part1(List<Command> commands) {
        var p = Position.ZERO;
        for (var cmd : commands) {
            p = p.update(cmd);
        }
        return p.product();
    }

    static int part2(List<Command> commands) {
        var p = new PositionWithAim(Position.ZERO, 0);
        for (var command : commands) {
            p = p.update(command);
        }
        return p.pos.product();
    }

    public static void main(String[] args) throws IOException {
        var lines = Files.readAllLines(Path.of("../input.txt"));
        var commands = lines.stream().map(Command::parse).collect(toList());
        System.out.println(part1(commands));
        System.out.println(part2(commands));
    }
}

File: 02/python/main.py
from typing import Tuple
from functools import reduce


def part1(steps):
    def sum_pairs(p1, p2):
        return (p1[0] + p2[0], p1[1] + p2[1])

    x, y = reduce(sum_pairs, steps)
    return x * y


def part2(steps):
    aim, x_pos, depth = 0, 0, 0
    for fwd, aim_delta in steps:
        if aim_delta:
            aim += aim_delta
        else:
            x_pos += fwd
            depth += aim * fwd
    return x_pos * depth


with open("../input.txt") as f:

    def parse_step(line) -> Tuple[int, int]:
        direction, magnitude = line.split()
        magnitude = int(magnitude)
        if direction == "up":
            return 0, -magnitude
        elif direction == "down":
            return 0, magnitude
        elif direction == "forward":
            return magnitude, 0
        else:
            raise ValueError()

    steps = [parse_step(line.strip()) for line in f]
    print(part1(steps))
    print(part2(steps))
